Resiliency: function even with missing status & TaskRun info for PipelineRun page

a-roberts opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes
For https://github.com/tektoncd/dashboard/issues/567

Without this change, every fifth(ish) rebuild would fail to render due to either being unable to find .steps from something undefined, or taskruns. With this change, add hardening (returning empty in some cases, or a small amount of data so we can render the page in others), and then rely on the websocket and new data coming in to update the components accordingly.

Tested with over 20 rebuilt runs in a row and noticing the components continue to be updated with new PipelineRun information accordingly.

Also includes a backend change to generate a more unique suffix (delegating to the k8s GenerateName feature).

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
Nope - sorry, desperate to get this in for 0.2.0! Thinking about adding...

- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

a-roberts commented:
Got a couple legit unit test failures so on the case

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -47,6 +47,15 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
+      return {};
+    }
+
+    if (!pipelineRun.status.taskRuns) {
```
let's combine these:

```js
if (!(pipelineRun.status && pipelineRun.status.taskRuns)) {
```

can we return more than just an empty object? e.g. at least `{ pipelineRun }`

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -65,11 +74,36 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
       return [];
     }
 
+    if (!pipelineRun.status) {
```
let's combine these:

```js
if (!(pipelineRun && pipelineRun.status && pipelineRun.status.taskRuns && tasks && taskRuns)) {
  return [];
}
```

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -65,11 +74,36 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
       status: { taskRuns: taskRunDetails }
     } = pipelineRun;
 
+    if (!pipelineRun.status) {
```
shouldn't these be caught by the checks above?

a-roberts commented:
Unit tests now fixed, displaying a lovely error message if it can't be loaded (verified by going to the PipelineRuns page for a non-existent PipelineRun I just deleted), linter errors addressed too

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -83,11 +117,26 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
+
+        let steps = '';
+
+        if (!task.spec || !taskRun.status) {
```
do you have a scenario that reliably gets us here? I'd like to see if there are other options as introducing fake data here can make this less flexible, especially when looking at how others might integrate the component

a-roberts replied:
Hmmmmmm, well I can reliably get taskRuns without a status coming in yeah, it's with five-ish rebuild runs in a row (you just keep rebuilding then clicking through to view that new one, then rebuild that, etc)

Same goes for the task steps, it's like the data just doesn't come in on time or we don't populate the store on time (and yet we do an async and wait on the response, so I'm guessing it's the websocket)...

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -190,6 +239,11 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
     }
     const taskRuns = this.loadTaskRuns(pipelineRun, taskRunNames);
+
+    if (taskRuns === []) {
```
`===` won't work here, `taskRuns.length` should be ok

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -190,6 +239,11 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
     const taskRuns = this.loadTaskRuns(pipelineRun, taskRunNames);
+
+    if (taskRuns === []) {
+      return <div>No task runs found for this PipelineRun</div>;
```
`TaskRuns`

also should be using react-intl here to externalise the string, either `<FormattedMessage ... />` or `intl.formatMessage(...)`

AlanGreene commented on packages/components/src/components/TaskTree/TaskTree.js:
```diff
@@ -23,9 +23,17 @@ class TaskTree extends Component {
   render() {
     const { selectedTaskId, taskRuns } = this.props;
+
+    if (!taskRuns) {
```
ideally we'd catch this higher up the tree, but if we want to render nothing we can just return `null`

a-roberts replied:
Weird, with this and the below (index.js change) I can reliably get `Uncaught TypeError: steps.map is not a function`, keeping this unchanged for now I think 🤔

AlanGreene commented on packages/utils/src/utils/index.js:
```diff
@@ -32,11 +32,17 @@ export function getErrorMessage(error) {
 
 export function selectedTaskRun(selectedTaskId, taskRuns = []) {
-  return taskRuns.find(run => run.id === selectedTaskId);
+  if (taskRuns) {
```
taskRuns should always be defined (we default to `[]`) here unless we're somehow passing `null` instead of `undefined`

if that's the case it should be caught before we get here
similar for `selectedTask` above, ideally we wouldn't get that far without tasks

AlanGreene commented on src/containers/PipelineRun/PipelineRun.js:
```diff
@@ -97,6 +97,21 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
   render() {
     const { error, intl, match, pipelineRun, tasks, taskRuns } = this.props;
+
+    if (!this.props.pipelineRun) {
```
no need for `this.props`, we've already deconstructed the props to extract `pipelineRun` so you can reference it directly here and below

AlanGreene replied:
missed this one? few of these:

```diff
-this.props.pipelineRun
+pipelineRun
```

a-roberts commented:
> Looks like we have quite a few weak spots, some of these changes will certainly help but I think it might be worth looking at another approach (separate from this PR) in the longer term.
> 
> Added some comments / suggestions / questions. Feel free to reach out on slack.

Yeah definitely, at this point it was a little "trial and error" to try and figure out all of the variables that aren't end up set for some reason. I'd like it if instead we continued to be a loading state while the remaining PipelineRun data makes it into the store

Will address the comments now, thanks for taking a look

a-roberts commented:
Ready for another look please @AlanGreene

AlanGreene commented on src/containers/PipelineRun/PipelineRun.js:
```diff
@@ -97,6 +97,21 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
+    }
+
+    if (!this.props.pipelineRun.status) {
+      pipelineRun.status = 'unknown';
```
didn't spot this earlier but looks like something fishy going on here?

`status = 'unknown'`
then immediately below
`status = somethingElse`

was this supposed to be updating `reason` or perhaps a `status` field somewhere else?

a-roberts replied:
Hmmmmmm, so what's there is intentional - if there's no status, set a status with no taskruns for that field

If there was actually a status anyway we enter the .status.taskRuns check and we provide an empty array of taskruns there

AlanGreene replied:
within this single conditional block you assign a value to status, then immediately reassign it

so the `status= 'unknown'` is redundant

a-roberts replied:
True, addressed and pushed, oops

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.js:
```diff
@@ -190,6 +228,16 @@ export /* istanbul ignore next */ class PipelineRunContainer extends Component {
+      return (
+        <FormattedMessage
+          id="dashboard.taskRun.noTaskRuns"
+          defaultMessage="No TaskRuns found for this PipelineRun yet..."
```
can we use an actual ellipsis here? `…`

AlanGreene commented on packages/components/src/components/PipelineRun/PipelineRun.test.js:
```diff
@@ -106,3 +106,54 @@ it('PipelineRunContainer handles error state', async () => {
   await waitForElement(() => getByText('Error loading PipelineRun'));
 });
+
+// A scenario exists (typically with rebuild) where all PipelineRun data
```
👍

a-roberts commented:
^^ fixed a few comments and added a note for a query

a-roberts commented:
Ready for another look please 😅

AlanGreene commented on jest.config.js:
```diff
@@ -23,7 +23,7 @@ module.exports = {
     global: {
       statements: 90,
-      branches: 90,
+      branches: 85,
```
@steveodonovan and I did some work earlier today to bring our coverage back up to a safer level. Can you revert this please? (I've verified your PR is fine locally with this at 90). It's too easy to let this stuff slip otherwise.

a-roberts replied:
Great, yeah will do! Pained me to do this, thanks

a-roberts replied:
Right I've rebased and noticed 90+ percent coverage across the board now on average
